graftm/hmmer.py

```diff
@@ -453,52 +453,35 @@ def alignment_correcter(self, alignment_file_list, output_file_name):
                         output_file.write(fasta_id)
                         output_file.write(fasta_seq)
 
-    def extract_orfs(self, input_path, orfm, orf_out_path):
+
+    def _extract_orfs(self, input_path, orfm, hit_readnames, output_path):
         '''
-        extract_orfs - Return a path to a FASTA file containing ORFs that hit 
-        the HMM (self.search_hmm)
+        Provide the read names of orfM orfs that hit, the nucleotide sequences
+        and the orfM command line. First writes the readnames to a tmp file,
+        to be used by fxtract to extract sequences. Then run command, which 
+        calls orfs, and pipes them to fxtarct, which extracts the reads of the 
+        hit orfs we're interested in. 
         
         Parameters
         ----------
         input_path : str
-            Path to input sequences in fasta nucleotide format
+            Path to input nucleotide sequences in FASTA format.
         orfm: obj
             graftm.OrfM object with parameters already set
-        orf_out_path
-            Path to output fasta file, containing amino acid ORFs
-        '''
-        raw_orf_path = tempfile.NamedTemporaryFile(prefix='orf_hmmsearch').name
-        orf_titles_path = tempfile.NamedTemporaryFile(prefix='orf_titles').name
-        
-        # Build the output_files
-        if len(self.search_hmm) >= 1:
-            output_table_list = [tempfile.NamedTemporaryFile(prefix='orf_hmmsearch').name for x in self.search_hmm]
-        else:
-            raise Exception("Programming error: expected 1 or more HMMs")
-
-        # Build ORF calling command orfs on the sequences
-        orfm_cmd = orfm.command_line()
-        cmd = '%s %s > %s' % (orfm_cmd, input_path, raw_orf_path)
-        
-        logging.debug("Running command: %s" % cmd)
-        subprocess.check_call(cmd, shell=True)
-        
-        cmd = 'cat %s' % raw_orf_path
-        searcher = HmmSearcher(1)
-        searcher.hmmsearch(cmd, self.search_hmm, output_table_list)
-        with open(orf_titles_path, 'w') as output:
-            reads = set(
-                       itertools.chain(
-                                       *[HMMreader(x).names() for x \
-                                         in output_table_list]
-                                       )
-                       )
-            [output.write(x + '\n') for x in reads]
-            
-        # Extract the reads using the titles.
-        cmd = 'fxtract -H -X -f %s %s > %s' % (orf_titles_path, raw_orf_path, orf_out_path)
-        logging.debug("Running command: %s" % cmd)
-        subprocess.check_call(cmd, shell=True)
+        hit_readnames : str
+            path to a file containin the readnames of hits to the HMM, one per 
+            line.
+        '''        
+        # Write hit readnames to file
+        with tempfile.NamedTemporaryFile(prefix='graftm_readnames') as orfm_readnames:
```

Comment:
something like this
```
In [12]: s = subprocess.Popen(['head','-n2'], stdin=subprocess.PIPE)

In [13]: s.communicate('1\n2\n3\n4\n')
1
2
Out[13]: (None, None)
```
I think you can use subprocess.check_output instead of Popen.